README.md

```diff
@@ -54,22 +54,40 @@ C:\vcpkg> vcpkg install wil:x64-windows
 Note that even though WIL is a header-only library, you still need to install the package for all architectures/platforms you wish to use it with. Otherwise, WIL won't be added to the include path for the missing architectures/platforms. Execute `vcpkg help triplet` for a list of available options.
 
 # Building/Testing
-To get started testing WIL, first make sure that you have a recent version of [Visual Studio](https://visualstudio.microsoft.com/downloads/) and the most recent [Windows SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk) installed. If you are doing
-any non-trivial work, also be sure to have a recent version of [Clang](http://releases.llvm.org/download.html) installed. Once everything is installed, open a VS
-native command window (e.g. "x64 Native Tools Command Prompt for VS 2022"). If you are familiar with CMake you can get started building normally. Otherwise, or if you prefer to skip all of the boilerplate, you can use one of the scripts in the [scripts](scripts) directory:
-```cmd
-C:\wil> scripts\init.cmd -c clang -g ninja -b debug
-```
-You can execute `init.cmd --help` for a summary of available options. The scripts use a common directory pattern of `build/$(compiler)$(arch)$(type)` for the build output root. E.g. `build/clang64debug` when using Clang as the compiler, x64 as the architecture, and Debug as the build type. It is this directory where you will want to build from. For example, if you initialized using the command above, you can build the tests like so:
+
+To get started contributing to WIL, first make sure that you have:
+
+* a recent version of [Visual Studio](https://visualstudio.microsoft.com/downloads/)
+* the most recent [Windows SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk)
+* [Nuget](https://www.nuget.org/downloads)
+
+If you are doing any non-trivial work, also be sure to have:
+
+* a recent version of [Clang](http://releases.llvm.org/download.html)
+ 
+Once everything is installed, open a VS native command window (e.g. `x64 Native Tools Command Prompt for VS 2022` \[_not_ `Developer Command Prompt for VS2022`]).
+
+* If you are familiar with CMake you can get started building normally.
+* Otherwise, or if you prefer to skip all of the boilerplate, you can use one of the scripts in the [scripts](scripts) directory:
+  ```cmd
+  C:\wil> scripts\init.cmd -c clang -g ninja -b debug
+  ```
+* To generate a Visual Studio solution with IntelliSense:
```

Comment:
Note that the `init.cmd` script invokes CMake with `CMAKE_EXPORT_COMPILE_COMMANDS=ON`, which will generate a `compile_commands.json` file which many editors - such as Visual Studio Code - recognize and use to provide code completion/suggestions/"error squiggles"/etc. This can also be specified manually if you invoke CMake yourself.

Perhaps it's just because it's in the context of a PR, but this kind of reads to me as Visual Studio being the _only_ way to get these features. I think the wording here is _okay_, but it might be worthwhile to at least document above that a `compile_commands.json` will be generated automatically when running the script.

Reply:
I actually posted this PR as a way of figuring that out :)

The only way I know how to get working IntelliSense is via Visual Studio---do you have reference on how to get it working in VS Code?

Reply:
I'll probably eat my words for saying this, but it kind of "just works." For example, if you just run the setup command with no arguments
```cmd
> scripts\setup.cmd
```
Then it'll set up the project to compile with 'Clang' as the compiler and 'Debug' as the build type, writing the build files to `build/clang64debug` (if it's a 64-bit tools window). Directly under this folder there will be a `compile_commands.json` file (`build/clang64debug/compile_commands.json`). VS Code should prompt you asking if you'd like to use this file to configure Intellisense:

![image](https://user-images.githubusercontent.com/40036384/226073288-b3a88735-2d8e-4602-8c64-1caa2ac4e502.png)

Assuming you select "yes" then VS Code will use that file to determine compiler command line arguments (such as C++ version, include paths, etc.) for each individual file. Note that this preference is saved in the `.vscode/c_cpp_properties.json` file. From there, you should get pretty comparable smart auto-complete, error squiggles, etc.

One semi-drawback is if a file is compiled as part of more than one project - as we do for the various "normal"/"noexcept"/etc. "flavors" of the tests - I'm unaware of how to "switch" between the different configurations. In practice, I've never really had a significant issue with that, though.

Reply:
Unfortunately, I spent at least an hour on this and was unable to get IntelliSense working in VS Code.

I installed the recommended C++ extensions (C++ and CMake Tools), but I never saw a prompt to consume the `compile_commands.json`. I saw a prompt to configure CMake, but that failed multiple times (once related to the `RelWithDebInfo` flag being missing(?)), and the auto-detected configurations did not match the `init.cmd` configurations.

I restructured the content a bit, introducing a clear place to add VS Code setup instructions, but I don't feel comfortable adding the setup instructions at this time. I'd love to work further on this, though.

Reply:
Hmm, that's interesting. I've never had that happen before across at least half a dozen machines (most of which having both Code and Code Insiders installed). The C/C++ extension is definitely necessary as that's the one doing the generation/intellisense/etc. You _should_ be able to set it up manually in the `c_cpp_properties.json` file with something like the following, though doing it manually has never been necessary for me.
```json
{
    "configurations": [
        {
            "name": "Win32",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [
                "_DEBUG",
                "UNICODE",
                "_UNICODE"
            ],
            "windowsSdkVersion": "10.0.22000.0",
            "compilerPath": "cl.exe",
            "cStandard": "c17",
            "intelliSenseMode": "windows-msvc-x64",
            "compileCommands": "${workspaceFolder}/build/clang64debug/compile_commands.json"
        }
    ],
    "version": 4
}
```

Reply:
This all assumes you've opened VSCode in "directory" view from the root folder. E.g.
```cmd
C:\wil> code .
```
If you have VSCode open before running CMake, then you _might_ need to reload the window to pick it up.

Reply:
OK, I may have gotten _somewhere_ with your configuration, but I think VS Code is just doing its best, ignoring your config:

---

First, some background:

* VS Code Insiders
* Running "open folder" in the root folder of the repo.
* C++ and CMAKE extensions installed (although I disabled the CMAKE extension to try some stuff)

---

Using your configuration, IntelliSense seems to work, but is unreliable---sometimes it dies, with no indication why, for a minute or 2. Ctrl-Shift-B offers multiple options, but all of them fail with something like this (`cl.exe` doesn't exist on my box, so that has a different error):

```
 * Executing task: C/C++: clang-cpp.exe build active file 

Starting build...
"C:\Program Files\LLVM\bin\clang-cpp.exe" -fcolor-diagnostics -fansi-escape-codes -g C:\Users\besto\Projects\wil\tests\WatcherTests.cpp -o C:\Users\besto\Projects\wil\tests\WatcherTests.exe
C:\Users\besto\Projects\wil\tests\WatcherTests.cpp:2:10: fatal error: 'wil/filesystem.h' file not found
#include <wil/filesystem.h>
         ^~~~~~~~~~~~~~~~~~
1 error generated.

Build finished with error(s).
```

F5 prompts to launch a debugger (GDB or Windows), but selecting either has no effect (dialog dismisses and nothing changes).

---

It looks like all of that came from the VS Code C++ extension, not from the c_cpp_properties.json.

Re-enabling the CMake extension, I do get _this_ prompt:

<img width="377" alt="Would you like to configure project" src="https://user-images.githubusercontent.com/432939/229604195-e6d53efc-cf73-458d-8df2-6ce75ef795b9.png">

I can select from these configurations, which do not match the configurations in init.cmd:

<img width="464" alt="image" src="https://user-images.githubusercontent.com/432939/229604550-67598d9e-0262-4079-8e47-331fc932ac98.png">

But none seem to work. Here's the output from configuration `Visual Studio Enterprise 2022 Release - x86_amd64`:

<details>
<summary>Build output</summary>

```
[main] Building folder: wil 
[visual-studio] Patch Windows SDK path from C:\Program Files (x86)\Windows Kits\10\bin\x86 to C:\Program Files (x86)\Windows Kits\10\bin\10.0.22000.0\x86 for C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Auxiliary\Build\vcvarsall.bat
[main] Configuring project: wil 
[proc] Executing command: "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -SC:/Users/besto/Projects/wil -Bc:/Users/besto/Projects/wil/build -G "Visual Studio 17 2022" -T host=x86 -A x64
[cmake] Not searching for unused variables given on the command line.
[cmake] -- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22624.
[cmake] -- The C compiler identification is MSVC 19.35.32216.1
[cmake] -- The CXX compiler identification is MSVC 19.35.32216.1
[cmake] -- Detecting C compiler ABI info
[cmake] -- Detecting C compiler ABI info - done
[cmake] -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.35.32215/bin/Hostx86/x64/cl.exe - skipped
[cmake] -- Detecting C compile features
[cmake] -- Detecting C compile features - done
[cmake] -- Detecting CXX compiler ABI info
[cmake] -- Detecting CXX compiler ABI info - done
[cmake] -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.35.32215/bin/Hostx86/x64/cl.exe - skipped
[cmake] -- Detecting CXX compile features
[cmake] -- Detecting CXX compile features - done
[cmake] Feeds used:
[cmake]   C:\Users\besto\.nuget\packages\
[cmake]   https://api.nuget.org/v3/index.json
[cmake] 
[cmake] Package "Microsoft.Windows.CppWinRT.2.0.221121.5" is already installed.
[cmake] CMake Error at tests/CMakeLists.txt:44 (if):
[cmake]   if given arguments:
[cmake] 
[cmake]     "STREQUAL" "RelWithDebInfo"
[cmake] 
[cmake]   Unknown arguments specified
[cmake] 
[cmake] 
[cmake] -- Configuring incomplete, errors occurred!
[cmake] See also "C:/Users/besto/Projects/wil/build/CMakeFiles/CMakeOutput.log".
[proc] The command: "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -SC:/Users/besto/Projects/wil -Bc:/Users/besto/Projects/wil/build -G "Visual Studio 17 2022" -T host=x86 -A x64 exited with code: 1
[visual-studio] Patch Windows SDK path from C:\Program Files (x86)\Windows Kits\10\bin\x86 to C:\Program Files (x86)\Windows Kits\10\bin\10.0.22000.0\x86 for C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Auxiliary\Build\vcvarsall.bat
```
</details>

---

I'd love to include VS Code instructions in this change, but does it make more sense to break this into a separate task and check the existing stuff in? The generated `.sln` works pretty well, and we clearly need to do more investigation for VS Code.

Reply:
1. I have _a_ CMake extension installed, but have never used it for anything more than syntax highlighting/auto-complete in CMakeLists.txt files. I've always done all my CMake invocation from the command line.
2. Building/debugging requires separate setup through VS Code mechanisms. E.g. my setup for _building_ that I did mostly manually. As you can see, I just set it up to invoke Ninja. Debugging/running requires similar setup, but I don't typically bother with it and use the command line:
```json
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Build Clang x64 Debug",
            "type": "shell",
            "command": "ninja",
            "options": {
                "cwd": "${workspaceRoot}/build/clang64debug"
            },
            "group": {
                "kind": "build",
                "isDefault": true
            }
        },
        {
            "label": "Build Clang x86 Debug",
            "type": "shell",
            "command": "ninja",
            "options": {
                "cwd": "${workspaceRoot}/build/clang32debug"
            },
            "group": {
                "kind": "build"
            }
        }
    ]
}
```
3. I think what you have written now is fine. The way it reads doesn't imply that VS is the _only_ way to get intellisense. Anyone familiar with working in VS Code should know how to set up their environment.

Reply:
Forgot to say that the JSON above is from `tasks.json`, which is a VS Code concept, not related to the C/C++ extension

Reply:
Gotchu---I have some familiarity with `tasks.json` on my own, so that makes sense. Your explanation also aligns with my experiences---I was assuming VS Code gave you build/debug automatically somehow, and I was surprised.

It would be lovely to get F5 working end-to-end in VS Code, long-term; we may want to write some build tasks and check them in. But for now, I'm going to skip a VS Code section.

And yes, I'm happy y'all recommended the rewrite; I didn't like the implication that VS was the only way to get IntelliSense.